### DISCUSSION TITLE
Default values in form do not update correctly after deleting entry

### DISCUSSION THREAD

**Discussion by RolandHesse** [neuefische/questions#775] [Web] [answered]
In our form to create a new and/or edit an existing snippet, there is the possibility to add and delete fields for entering web-links. Here is the problem:

For instance, we have three fields for links: www.address1.com, www.address2.com, and www.address3.com. When clicking on the button next to the second field, what happens is that one field disappears (as intended), however, the remaining show incorrect addresses, i.e., www.address1.com and www.address2.com (when the desired result would be www.address1.com and www.address3.com).

Both the state which stores the entered URLs and the data sent to and received from the data-base are correctly updated.

The current state of our code can be found here: https://github.com/PeetThesing/capstone-project_snippet-manager/pull/41

Any suggestions welcome! 🙏

**Comment by doemser** [comment, marked as answer]
Your challenge revolves around using the index to handle essential program logic. You should never use the index for that.

I coded you a small prototype that handles your challenge isolated, so you have a reference and know where to go:

https://codesandbox.io/s/upbeat-thunder-6mn2tn?file=/src/App.js

Also you are using document.getElementByID which is an anti pattern in react, I solved it in this case with autoFocus.

Hope that helps!

**Reply by NadjaKanunnikow** [reply]
Hi there, thank you very much for your comment. Taking the lead from your prototype, everything seems to work now. 🙏 
Out of pure curiosity we would still like to know why we are not supposed to use the index. Obviously it didn't work, but we don't understand why. Maybe you could elaborate?

**Reply by doemser** [reply]
Using indices for array manipulation in React, such as adding or deleting items, can lead to several issues, especially if you’re managing state in a non-immutable way or using indices for other logic beyond simply tracking the position of an item. Here’s why you should be careful with this approach:

Immutability:
In React, it’s important to treat state as immutable. Directly modifying the state (e.g., using array.splice(index, 1) to remove an item) can lead to unpredictable component behavior since React state updates may be asynchronous and such mutations won't trigger re-renders properly. Always use methods like setState with a new array that has the items added or removed.

Synchronization Issues:
If you’re using indices for logic (such as deleting an item at a specific index), any change in the array (like sorting, adding, or removing elements) can cause a mismatch between the component’s state and what the indices represent. This can result in the wrong item being manipulated.

Non-Unique Identifiers:
When you delete an item from the middle of an array, the indices of all subsequent items change. If you were relying on those indices for identification purposes (like keys in a list), this could lead to issues similar to using indices as keys—incorrect items might be updated or removed.

Unintended Side Effects:
If other components or logic depend on the current index of an item, removing or adding items can cause a ripple effect of unintended updates elsewhere in your application because the reference index has changed.
